File: src/main/java/com/joraph/schema/Schema.java
package com.joraph.schema;

import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

import kotlin.jvm.functions.Function1;

/**
 * A schema.
 */
public class Schema {

	public static final Comparator<Class<?>> CLASS_COMPARATOR
		= new Comparator<Class<?>>() {
			@Override
			public int compare(Class<?> o1, Class<?> o2) {
				return o1.getName().compareTo(o2.getName());
			}
		};

	private boolean validated = false;
	private final Map<Class<?>, EntityDescriptor<?>> entityDescriptors = new HashMap<>();

	private void dirty() {
		this.validated = false;
	}

	/**
	 * @return whether the schema has been validated
	 */
	public boolean isValidated() {
		return validated;
	}

	/**
	 * Asserts that the schema has been validated.
	 */
	public void assertValidated() {
		// TODO remove this method and just validate at the appropriate call sites
		if (!isValidated()) {
			throw new IllegalStateException("Schema not validated");
		}
	}

	/**
	 * Returns the graph type key for the given entity class.
	 * @param entityClass the entity class
	 * @return the key
	 */
	public Class<?> getGraphTypeKey(Class<?> entityClass) {
		return Optional.ofNullable(entityDescriptors.get(entityClass))
				.map(EntityDescriptor::getGraphKey)
				.map(Class.class::cast)
				.orElse(entityClass);
	}

	/**
	 * @param entityClass the entity class
	 * @return the descriptor for the class, or null if no descriptor exists for it
	 */
	public Set<EntityDescriptor<?>> getEntityDescriptors(Class<?> entityClass) {
		return this.entityDescriptors.values().stream()
				.filter((ed) -> ed.getEntityClass().equals(entityClass)
					|| ed.getGraphKey().equals(entityClass))
				.collect(Collectors.toSet());
	}

	/**
	 * Adds an entity descriptor to the schema.
	 * @param entityDescriptor the entity descriptor to add
	 */
	public void addEntityDescriptor(EntityDescriptor<?> entityDescriptor) {
		this.dirty();
		this.entityDescriptors.put(entityDescriptor.getEntityClass(), entityDescriptor);
	}

	/**
	 * Wraps a class in an {@link com.joraph.schema.EntityDescriptor} and then
	 * adds it to the schema.
	 * @param entityClass the entity class
	 * @return the {@link com.joraph.schema.EntityDescriptor} created as a wrapper and
	 * added to the schema
	 */
	public <T> EntityDescriptor<T> addEntityDescriptor(Class<T> entityClass) {
		this.dirty();
		EntityDescriptor<T> ret = new EntityDescriptor<>(entityClass);
		addEntityDescriptor(ret);
		return ret;
	}

	/**
	 * Describes the foreign keys from one class to another, if any exist.
	 * @param fromEntityClass the from entity class
	 * @param toEntityClass the to entity class
	 * @return all of the foreign keys which exist from one entity class to
	 *         another or an empty collection if no such relationships exist
	 */
	public Collection<ForeignKey<?, ?>> describeForeignKeys(Class<?> fromEntityClass, Class<?> toEntityClass) {
		return describeForeignKeysFrom(fromEntityClass).stream()
				.filter((fk) -> fk.getForeignEntity().equals(toEntityClass))
				.collect(Collectors.toList());
	}

	/**
	 * Describes the foreign keys configured for a specified class.
	 * @param entityClass the entity class
	 * @return the foreign keys configured for that class
	 */
	public Collection<ForeignKey<?, ?>> describeForeignKeysFrom(Class<?> entityClass) {
		return getEntityDescriptors(entityClass).stream()
				.flatMap((entityDescriptor) ->
						entityDescriptor.getForeignKeys().values().stream())
				.collect(Collectors.toList());
	}

	/**
	 * Describes the foreign keys coming into a given entity class.
	 * @param toEntityClass the entity class
	 * @return foreign keys that point to an entity class
	 */
	public Collection<ForeignKey<?, ?>> describeForeignKeysTo(Class<?> toEntityClass) {
		return entityDescriptors.values().stream()
				.map(EntityDescriptor::getForeignKeys)
				.map(Map::values)
				.flatMap(Collection::stream)
				.filter((fk) -> fk.getForeignEntity().equals(toEntityClass))
				.collect(Collectors.toList());
	}

	/**
	 * Describes all of the foreign keys that have been configured.
	 * @return all of the configured foreign keys
	 */
	public Collection<ForeignKey<?, ?>> describeForeignKeys() {
		return entityDescriptors.values().stream()
				.map(EntityDescriptor::getForeignKeys)
				.map(Map::values)
				.flatMap(Collection::stream)
				.collect(Collectors.toList());
	}

	/**
	 * Describes the given entity class.
	 */
	public Set<Class<?>> describeEntities() {
		this.assertValidated();
		return Collections.unmodifiableSet(entityDescriptors.keySet());
	}

	/**
	 * Validates the {@link Schema}.
	 */
	public Schema validate() {

		// check each entity descriptor
		for (Entry<Class<?>, EntityDescriptor<?>> entry : entityDescriptors.entrySet()) {
			EntityDescriptor<?> ed = entry.getValue();

			// check pk
			ed.getPrimaryKey();

			// check FKs
			Map<Function1<?, ?>, ForeignKey<?, ?>> fks = ed.getForeignKeys();
			for (Entry<Function1<?, ?>, ForeignKey<?, ?>> fkEntry : fks.entrySet()) {
				ForeignKey<?, ?> fk = fkEntry.getValue();
				entityDescriptors.values().stream()
						.filter((d) -> d.getEntityClass().equals(fk.getForeignEntity())
								|| d.getGraphKey().equals(fk.getForeignEntity()))
						.findAny()
						.orElseThrow(() -> new UnknownFKException(ed.getEntityClass(), fk));
			}
		}

		// good to go
		this.validated = true;
		return this;
	}

}

File: src/main/kotlin/com/joraph/schema/Schema.kt
package com.joraph.schema

/**
 * The Schema holds all of the meta-data about an object graph, that is all of it's entities and their
 * primary and foreign keys.
 */
class Schema {

    var isValidated = false
        private set

    private val entityDescriptors: MutableMap<Class<*>, EntityDescriptor<*>> = mutableMapOf()

    /**
     * Asserts that the schema has been validated.
     */
    fun assertValidated() {
        check(isValidated) { "Schema not validated" }
    }

    /**
     * Returns the graph type key for the given entity class. This is the key that
     * will be used to store the given entity in the [com.joraph.ObjectGraph]. This is
     * useful for polymorphic types that need to be addressed as the same super type.
     */
    fun getGraphTypeKey(entityClass: Class<*>): Class<*> {
        val desc = entityDescriptors[entityClass]
            ?: return entityClass
        return desc.graphKey
    }

    /**
     * Returns all of the [EntityDescriptor] configured for the given entity.
     */
    fun getEntityDescriptors(entityClass: Class<*>): Set<EntityDescriptor<*>> {
        return entityDescriptors.values
            .filter { it.entityClass == entityClass || it.graphKey == entityClass }
            .toSet()
    }

    /**
     * Adds an [EntityDescriptor] to the schema.
     */
    fun addEntityDescriptor(entityDescriptor: EntityDescriptor<*>) {
        isValidated = false
        entityDescriptors[entityDescriptor.entityClass] = entityDescriptor
    }

    /**
     * Adds to the schema and returns an empty [EntityDescriptor] for the given entity.
     */
    fun <T> addEntityDescriptor(entityClass: Class<T>): EntityDescriptor<T> {
        isValidated = false
        val ret = EntityDescriptor(entityClass)
        addEntityDescriptor(ret)
        return ret
    }

    /**
     * Returns a [List] of [ForeignKey]s defined from the [fromEntityClass] to the [toEntityClass].
     */
    fun describeForeignKeys(fromEntityClass: Class<*>, toEntityClass: Class<*>): List<ForeignKey<*, *>> {
        return describeForeignKeysFrom(fromEntityClass)
            .filter { it.foreignEntity == toEntityClass }
            .toList()
    }

    /**
     * Returns a [List] of [ForeignKey]s defined on the [entityClass].
     */
    fun describeForeignKeysFrom(entityClass: Class<*>): List<ForeignKey<*, *>> {
        return getEntityDescriptors(entityClass)
            .flatMap { it.foreignKeys }
    }

    /**
     * Returns a [List] of [ForeignKey]s defined on other entities to the [entityClass].
     */
    fun describeForeignKeysTo(entityClass: Class<*>): List<ForeignKey<*, *>> {
        return entityDescriptors.values
            .flatMap { it.foreignKeys }
            .filter { it.foreignEntity == entityClass }
    }

    /**
     * Returns a [List] of all [ForeignKey]s defined.
     */
    fun describeForeignKeys(): List<ForeignKey<*, *>> {
        return entityDescriptors.values
            .flatMap { it.foreignKeys }
    }

    /**
     * Returns a [List] of all entities defined.
     */
    fun describeEntities(): Set<Class<*>> {
        assertValidated()
        return entityDescriptors.keys
    }

    /**
     * Validates the [Schema], throwing a [com.joraph.JoraphException] on error.
     */
    fun validate(): Schema {

        // check each entity descriptor
        for ((_, ed) in entityDescriptors) {

            // check pk
            ed.primaryKey

            // check FKs
            val fks: Set<ForeignKey<*, *>> = ed.foreignKeys
            for (fk in fks) {
                entityDescriptors.values
                    .find {  it.entityClass == fk.foreignEntity || it.graphKey == fk.foreignEntity }
                    ?: throw UnknownFKException(ed.entityClass, fk)
            }
        }

        // good to go
        isValidated = true
        return this
    }

}
